orca-echo/src/main/groovy/com/netflix/spinnaker/orca/echo/tasks/PageApplicationOwnerTask.groovy

```diff
@@ -42,32 +44,54 @@ class PageApplicationOwnerTask implements RetryableTask {
   @Autowired(required = false)
   Front50Service front50Service
 
+  private final ObjectMapper mapper = new ObjectMapper()
+
   @Override
   TaskResult execute(Stage stage) {
     if (!front50Service) {
       throw new UnsupportedOperationException("Front50 is not enabled, no way to fetch pager duty. Fix this by setting front50.enabled: true");
     }
-    def application = stage.context.application as String
-    def applicationPagerDutyKey = fetchApplicationPagerDutyKey(application)
-    if (!applicationPagerDutyKey) {
-      throw new IllegalStateException("${application} does not have a pager duty service key!")
+    def applications = stage.context.applications as List<String>
+    def keys = stage.context.keys as List<String>
+    def pagerDutyKeys = []
+
+    if (keys) {
+      // Add all arbitrary keys to the pager duty key list
+      pagerDutyKeys.addAll(keys)
     }
 
+    // Add applications to the pager duty key list if they have a valid key
+    if (applications && applications.size() > 0) {
+      def invalidApplications = []
+      applications.each { application ->
+        def applicationPagerDutyKey = fetchApplicationPagerDutyKey(application)
+        if (!applicationPagerDutyKey) {
+          invalidApplications.push(application)
+        } else {
+          pagerDutyKeys.push(applicationPagerDutyKey)
+        }
+      }
+
+      if (invalidApplications.size() > 0) {
```

Comment:
`if (invalidApplications)`

Q: if one of the applications doesn't have a pager duty key, do you want to fail everything and not send to other valid applications?

Maybe a moot point and something that won't really happen in reality?

Reply:
True, very good catch. I'll fix it to not fail entirely if one does not have a pd key. Very nice.

Reply:
Actually, I'll figure this out in another iteration. Moot point for our specific use case, but something very important to figure out....